docs/source/usage.rst

```diff
@@ -339,58 +380,112 @@ Input YAML file:
-    >>> print(cfg.database.password)
-    secret
+    >>> cfg.database.password
+    'abc123'
+
+Environment variables are parsed when they are recognized as valid quantities that
+may be evaluated (e.g., int, float, dict, list):
+
+.. doctest::
+
+    >>> cfg = OmegaConf.create({
+    ...       'database': {'password': '${env:DB_PASSWORD,abc123}',
+    ...                    'user': 'someuser',
+    ...                    'port': '${env:DB_PORT,3306}',
+    ...                    'nodes': '${env:DB_NODES,[]}'}
+    ... })
+    >>> os.environ["DB_PORT"] = '3308'
+    >>> cfg.database.port  # converted to int
+    3308
+    >>> os.environ["DB_NODES"] = '[host1, host2, host3]'
+    >>> cfg.database.nodes  # converted to list
+    ['host1', 'host2', 'host3']
+    >>> os.environ["DB_PASSWORD"] = 'a%#@~{}$*&^?/<'
+    >>> cfg.database.password  # kept as a string
+    'a%#@~{}$*&^?/<'
+
 
 Custom interpolations
 ^^^^^^^^^^^^^^^^^^^^^
+
 You can add additional interpolation types using custom resolvers.
-This example creates a resolver that adds 10 the the given value.
+The example below creates a resolver that adds 10 to the given value.
 
 .. doctest::
 
-    >>> OmegaConf.register_resolver("plus_10", lambda x: int(x) + 10)
+    >>> OmegaConf.new_register_resolver("plus_10", lambda x: x + 10)
     >>> c = OmegaConf.create({'key': '${plus_10:990}'})
     >>> c.key
     1000
 
 
 Custom resolvers support variadic argument lists in the form of a comma separated list of zero or more values.
 Whitespaces are stripped from both ends of each value ("foo,bar" is the same as "foo, bar ").
-You can use literal commas and spaces anywhere by escaping (:code:`\,` and :code:`\ `).
+You can use literal commas and spaces anywhere by escaping (:code:`\,` and :code:`\ `), or
+simply use quotes to bypass character limitations in strings.
 
 .. doctest::
 
-    >>> OmegaConf.register_resolver("concat", lambda x,y: x+y)
+    >>> OmegaConf.new_register_resolver("concat", lambda x, y: x+y)
     >>> c = OmegaConf.create({
     ...     'key1': '${concat:Hello,World}',
     ...     'key_trimmed': '${concat:Hello , World}',
     ...     'escape_whitespace': '${concat:Hello,\ World}',
+    ...     'quoted': '${concat:"Hello,", " World"}',
     ... })
     >>> c.key1
     'HelloWorld'
     >>> c.key_trimmed
     'HelloWorld'
     >>> c.escape_whitespace
     'Hello World'
+    >>> c.quoted
+    'Hello, World'
+
+You can take advantage of nested interpolations to perform custom operations over variables:
+
+.. doctest::
+
+    >>> OmegaConf.new_register_resolver("plus", lambda x, y: x + y)
+    >>> c = OmegaConf.create({"a": 1,
+    ...                       "b": 2,
+    ...                       "a_plus_b": "${plus:${a},${b}}"})
+    >>> c.a_plus_b
+    3
+
+By default a custom resolver's output is cached, so that when it is called with the same
+inputs we always return the same value. This behavior may be disabled by setting ``use_cache=False``:
```

Comment:
No, the caching is performed when the resolver is called, and at this time `${a}` and `${b}` are already resolved, so the cache is based on the actual values of a & b (not on the interpolation strings).

Reply:
What if we used the literal values in the interpolation as the cache keys?
I understand that it will may require changes to how things are done with the parsing now.

I think this would make the behavior much more predictable.
Let's chat about the cache behavior again to make the final decision.

Reply:
> What if we used the literal values in the interpolation as the cache keys?

This isn't an option I had considered as so far I had always seen the cache meant to be used only for either (i) non-deterministic resolvers, or (ii) expensive resolvers. I assumed that for a determnistic resolvers, we would have liked the behavior to remain the same whether it is cached or not.

I create #526 to discuss the cache behavior.